Improve accepted JSON values in JsonObjectParser

cchepelov opened the pull request:
This PR fixes #25673 by making the JsonObjectParser framing engine able to also accept:

- arrays of scalar Json values (strings, booleans, numbers)
- linebreak or comma-separated scalar Json values 

The tests are expanded to reflect this improvement.

In terms of implementation, the increased complexity led me to extract the state machine logic (formerly in ```def proceed```) into independent private state objects (which are stored within the companion in order to avoid creating tons of instances at each new stream). The JsonObjectParser will now behave very differently according to the following keys:
  - if the first non-whitespace character is a ```[```, then the parser assumes the "JSON Array" encoding and will enforce a single value per item, separated by commas, and optionally terminated with a single ```]``` (it will not complain if that is missing).
  - otherwise, it runs in "separator-separated mode". This mode is subdivided in comma or linebreak-separated mode, depending in which separator is found after the first value. 

Furthermore, it was discovered as part of running the JMH Benchmark that the previous implementation of JsonObjectParser was able to accept streams of unseparated JSON *Object* values, i.e. it did correctly frame the following stream:
```json
  {"foo":1}{"foo":2}{"foo":3}
```
even though there is neither a comma or a line break between objects.
The tests are further expanded to reflect this. All three of "yet undetermined separator-separated", "comma-separated" and "linebreak-separated" modes accept the above sequence.

In terms of performance, at at **-wi 5 -i 5** iterations:

**Before**, (https://github.com/akka/akka/commit/16438fcb860f932aec4187c27053884dcf2d9fe5)
```text
Benchmark                                     Mode  Cnt        Score        Error  Units
JsonFramingBenchmark.counting_1              thrpt    5  2084711,811 ± 207865,821  ops/s
JsonFramingBenchmark.counting_long_document  thrpt    5      186,252 ±    116,947  ops/s
JsonFramingBenchmark.counting_offer_5        thrpt    5  2306649,473 ± 471682,266  ops/s
```
**After**:
```text
Benchmark                                     Mode  Cnt        Score        Error  Units
JsonFramingBenchmark.counting_1              thrpt    5  3418384,183 ± 481082,430  ops/s
JsonFramingBenchmark.counting_long_document  thrpt    5      728,951 ±     30,755  ops/s
JsonFramingBenchmark.counting_offer_5        thrpt    5  2772735,430 ± 321462,765  ops/s
```

When running at a high amount of iterations, it seems something strange kicks in (running with **-wi 30 -i 30**):
**Before**
```text
JsonFramingBenchmark.counting_1              thrpt   30  2565875,250 ± 24452,637  ops/s
JsonFramingBenchmark.counting_long_document  thrpt   30      190,048 ±     2,777  ops/s
JsonFramingBenchmark.counting_offer_5        thrpt   30  2188289,936 ± 32093,784  ops/s
```

**After**
```text
Benchmark                                     Mode  Cnt        Score       Error  Units
JsonFramingBenchmark.counting_1              thrpt   30  2409937,623 ± 10269,970  ops/s
JsonFramingBenchmark.counting_long_document  thrpt   30      700,063 ±    15,331  ops/s
JsonFramingBenchmark.counting_offer_5        thrpt   30  2078812,214 ± 45579,564  ops/s
```
There is a noticeable drop in throughput for ```counting_1``` and ```counting_offer_5``` after the 20th warm-up iteration. It is unclear to me whether JsonObjectParser, thermal CPU throttling or the test fixture itself is the culprit (the nominal CPU frequency appeared to stay put at 3509 MHz throughout the test runs)
```text
# JMH version: 1.21
# VM version: JDK 1.8.0_171, OpenJDK 64-Bit Server VM, 25.171-b11
# scala 2.12.6
```

Commit:
JsonFramingSpec: Add multiple test cases of legal JSON that is currently rejected

Commit:
Refine test JSON framing cases

Commit:
JsonObjectParser: support parsing naked scalars (WIP: except strings)

Commit:
JsonObjectParser: support parsing string scalars (as arrays, newline or comma-separated streams)

Commit:
JsonObjectParser: rebuilt from the ground up to better support corner cases

Commit:
Remove println's

Commit:
Recycle old ContainerStackLevel structures as we leave and enter JSON scopes

Commit:
(code formatting)

Commit:
New tests: some streams used in JMH benchmarks fail, why?

Commit:
…Tolerate consecutive JSON objects without separators

Commit:
tidy up a bit the interface between JsonObjectParser & its components

Commit:
use numeric constants rather than 'c'.toByte, which scalac doesn't inline

Commit:
Enable use of tail recursion when remaining in the same parser state

Commit:
avoid internal buffer compactions

johanandren commented:
OK TO TEST

Commit:
additional mima excludes (from an @InternalApi companion)

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/JsonObjectParser.scala:
```diff
@@ -36,6 +37,469 @@ import scala.annotation.switch
+          }
+        case _ ⇒
+          throw new FramingException(s"Invalid JSON encountered at position [${pp.pos}] of [${pp.buffer}] after closed JSON-style array")
+      }
```
Just do an `if-else` instead of pat match since there is just two branches on a boolean condition

cchepelov replied:
Doing this. 

Note that this came from using a kind of "cookie-cutter" approach to managing the conditions, and might be the remnants of additional conditions at this specific place. 
While the bytecode is not strictly identical, there should be few hits in practice anyway (whitespace after the end of an array-formatted stream)

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/JsonObjectParser.scala:
```diff
@@ -36,6 +37,469 @@ import scala.annotation.switch
+            proceed(pp.nextInput, pp)
+          }
+
+        case _ if input == containerEnd ⇒
```
```case `containerEnd` => ```

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/JsonObjectParser.scala:
```diff
@@ -36,6 +37,469 @@ import scala.annotation.switch
+          pp.enterState(InNaked(this, pp))
+
+        //case _ =>
+        //  throw new FramingException(s"Invalid JSON encountered at position [${pp.pos}] of [${pp.buffer}]")
```
Remove

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/JsonObjectParser.scala:
```diff
@@ -36,6 +37,469 @@ import scala.annotation.switch
+    }
+
+    private object InArray extends InContainerBase(SquareBraceEnd) with LeaveContainerOnExit {
+      override def toString: String = "InArray"
```
If you make these case objects they would get the toString "for free"

johanandren replied:
Doesn't matter much but for the record I think you missed that I said **case** object - will be exactly like this, not a FQCN

cchepelov replied:
I did miss you said **case** object indeed! just pushed a correction.

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/JsonObjectParser.scala:
```diff
@@ -36,6 +37,469 @@ import scala.annotation.switch
+      def beforeNextItem: ParserState
+
+      final override def proceed(input: Byte, pp: JsonObjectParser): Unit = input match {
+        case _ if input == separator ⇒
```
```case `separator` => ```

cchepelov replied:
actually… `stable identifier required, but SeparatorSeparatedAfterElement.this.separator found.`

cchepelov replied:
(found a way around this)

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/JsonObjectParser.scala:
```diff
@@ -36,6 +37,469 @@ import scala.annotation.switch
+        case _ if isWhitespace(input) ⇒
+          pp.skip()
+
+        case _ if !isWhitespace(input) ⇒
```
Guard not needed here, the byte is either whitespace and got caught by the above clause or it is not and then it will end up here.

cchepelov replied:
stylistic issue. I tend to like this for clarity of symmetry. DRY is a fair opposite. Removing.

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/JsonObjectParser.scala:
```diff
@@ -36,6 +37,469 @@ import scala.annotation.switch
+
+      @tailrec
+      final override def proceed(input: Byte, pp: JsonObjectParser): Unit = {
+        if (input == SquareBraceStart) {
```
use pattern matching here since you do in all the other places

cchepelov replied:
good point ;)

johanandren reviewed (commented):
Looks promising. Left some comments.

One worry is that it looks like it allocates quite a few more objects than the original parser, would be good if you could benchmark it for memory pressure as well as throughput. (`-perf gc` for jmh).

cchepelov commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/JsonFramingSpec.scala:
```diff
@@ -519,5 +867,155 @@ class JsonFramingSpec extends AkkaSpec {
+          |[ "abc", { "def": "ghi" } ]
+          |""".stripMargin
+
+      /* we can't reliably support this while also automatically supporting the "JSON Array of things" style */
```
Correct. Some of the additional tests here pin the observed prior behaviour as a means to document implicits.

cchepelov commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/JsonFramingSpec.scala:
```diff
@@ -519,5 +867,155 @@ class JsonFramingSpec extends AkkaSpec {
+    }
+  }
+
+  "running tests used for benchmarking" should {
```
The benchmark was relying on specific implicit and undocumented behaviour. Copying this into the test section documents and pins the implicit undocumented behaviour as my first attempts changed the implementation in a way that broke the benchmark, and by implication risked breaking deployed user code.

Would you like me to add a comment explaining why the apparent duplication?

johanandren replied:
Nah, that's fine then, but adding a little comment stating this would be nice.

cchepelov commented:
Thanks for the review @johanandren and apologies for the weeks in a tunnel.

Regarding the GC, will benchmark. I tried hard to avoid creating objects at runtime (there are a number of objects that get created as part of the stream's init, and a few that can get created as part of the state stack, but there's unconventional recycling for this). 

Will report, I understand this is an important metric to reach a decision.

johanandren commented:
No worries, we'll be fairly busy in the team the coming weeks as well, so don't feel any pressure to hurry up from our side.

Commit:
cleanups following @johanandren's review https://github.com/akka/akka/pull/25674#pullrequestreview-158939301

cchepelov commented:
Post-review, revisiting performance:

master (d9a41f5e237a841b7f15e198fb5c22673b6ed1cf):
```text
[info] Benchmark                                     Mode  Cnt        Score        Error  Units
[info] JsonFramingBenchmark.counting_1              thrpt    5  2876586,577 ± 104206,810  ops/s
[info] JsonFramingBenchmark.counting_long_document  thrpt    5      216,529 ±      4,662  ops/s
[info] JsonFramingBenchmark.counting_offer_5        thrpt    5  2436713,062 ±  85255,671  ops/s
```

Initial pull request (641961ddcf8a7bfd46cee6c06722573c76b485b4)
```text
[info] Benchmark                                     Mode  Cnt        Score        Error  Units
[info] JsonFramingBenchmark.counting_1              thrpt    5  3191640,194 ± 787917,209  ops/s
[info] JsonFramingBenchmark.counting_long_document  thrpt    5      734,874 ±    128,198  ops/s
[info] JsonFramingBenchmark.counting_offer_5        thrpt    5  2754103,799 ± 384199,265  ops/s
```

Updated branch (bb456bd7d2c9b9524a35dbfe7c463e7e8967a7d6), today:
```text
[info] Benchmark                                     Mode  Cnt        Score        Error  Units
[info] JsonFramingBenchmark.counting_1              thrpt    5  3438107,284 ± 348074,002  ops/s
[info] JsonFramingBenchmark.counting_long_document  thrpt    5      750,122 ±     17,825  ops/s
[info] JsonFramingBenchmark.counting_offer_5        thrpt    5  2794009,199 ± 285732,751  ops/s
```

There seems to be a minute throughput improvement, possibly due to the improvement you suggested in patmat. I'll investigate whether it's worth expanding ` case _ if isWhitespace(input) ⇒` into `case Space|Linebreak|Linebreak2|Tab ⇒` despite the obvious maintenance drawback. 

Regarding the allocation question:
master (d9a41f5e237a841b7f15e198fb5c22673b6ed1cf):
```text
[info] JsonFramingBenchmark.counting_1                                     thrpt    5  2624091,126 ± 1219347,872   ops/s
[info] JsonFramingBenchmark.counting_1:·gc.alloc.rate                      thrpt    5      133,462 ±      62,014  MB/sec
[info] JsonFramingBenchmark.counting_1:·gc.alloc.rate.norm                 thrpt    5       56,000 ±       0,001    B/op
[info] JsonFramingBenchmark.counting_long_document                         thrpt    5      210,822 ±      13,792   ops/s
[info] JsonFramingBenchmark.counting_long_document:·gc.alloc.rate          thrpt    5        0,011 ±       0,001  MB/sec
[info] JsonFramingBenchmark.counting_long_document:·gc.alloc.rate.norm     thrpt    5       56,208 ±       0,015    B/op
[info] JsonFramingBenchmark.counting_offer_5                               thrpt    5  2477899,025 ±  115321,706   ops/s
[info] JsonFramingBenchmark.counting_offer_5:·gc.alloc.rate                thrpt    5      608,525 ±      28,318  MB/sec
[info] JsonFramingBenchmark.counting_offer_5:·gc.alloc.rate.norm           thrpt    5      270,400 ±       0,001    B/op
```

Updated branch (bb456bd7d2c9b9524a35dbfe7c463e7e8967a7d6), today:
```text
[info] JsonFramingBenchmark.counting_1                                         thrpt    5  3636627,358 ± 409547,095   ops/s
[info] JsonFramingBenchmark.counting_1:·gc.alloc.rate                          thrpt    5      132,115 ±     14,873  MB/sec
[info] JsonFramingBenchmark.counting_1:·gc.alloc.rate.norm                     thrpt    5       40,000 ±      0,001    B/op
[info] JsonFramingBenchmark.counting_long_document                             thrpt    5      764,573 ±     13,482   ops/s
[info] JsonFramingBenchmark.counting_long_document:·gc.alloc.rate              thrpt    5        0,028 ±      0,004  MB/sec
[info] JsonFramingBenchmark.counting_long_document:·gc.alloc.rate.norm         thrpt    5       40,700 ±      5,524    B/op
[info] JsonFramingBenchmark.counting_offer_5                                   thrpt    5  2942200,995 ± 102472,910   ops/s
[info] JsonFramingBenchmark.counting_offer_5:·gc.alloc.rate                    thrpt    5      209,497 ±      7,291  MB/sec
[info] JsonFramingBenchmark.counting_offer_5:·gc.alloc.rate.norm               thrpt    5       78,400 ±      0,001    B/op
```

johanandren commented:
Alloc looks good, thanks for double checking that.

Make sure to warmup until the throughput stabilizes and then probably run a few more than 5 iterations for the actual benchmark, just to make sure we're not seeing pre-jit performance etc in there.

cchepelov commented:
> There seems to be a minute throughput improvement, possibly due to the improvement you suggested in patmat. I'll investigate whether it's worth expanding `case _ if isWhitespace(input) ⇒` into `case Space|Linebreak|Linebreak2|Tab ⇒` despite the obvious maintenance drawback.

Not getting really obviously better numbers: 
```text
[info] Benchmark                                     Mode  Cnt        Score        Error  Units
[info] JsonFramingBenchmark.counting_1              thrpt    5  3438107,284 ± 348074,002  ops/s     → 3478521,210 ± 266340,707  ops/s
[info] JsonFramingBenchmark.counting_long_document  thrpt    5      750,122 ±     17,825  ops/s     → 758,564 ±     37,204  ops/s
[info] JsonFramingBenchmark.counting_offer_5        thrpt    5  2794009,199 ± 285732,751  ops/s     → 2849088,721 ± 277487,640  ops/s
```

while the average score seem slightly better in all cases, this is still quite within the last push metrics error intervals. Probably not worth the loss in code legibility; I'll leave it as is.

cchepelov commented:
Running with 30 warm-up & 30 iterations yields weird results (as in the initial report):

Master d9a41f5e237a841b7f15e198fb5c22673b6ed1cf at -i 30 -wi 30:
```text
[info] JsonFramingBenchmark.counting_1              thrpt   30  2655804,989 ± 60613,486  ops/s
[info] JsonFramingBenchmark.counting_long_document  thrpt   30      201,483 ±     5,793  ops/s
[info] JsonFramingBenchmark.counting_offer_5        thrpt   30  2285532,218 ± 59965,221  ops/s
```

Updated branch bb456bd7d2c9b9524a35dbfe7c463e7e8967a7d6 at -i 30 -wi 30:
```text
[info] JsonFramingBenchmark.counting_1              thrpt   30  2488869,333 ± 36421,754  ops/s
[info] JsonFramingBenchmark.counting_long_document  thrpt   30      720,803 ±    12,887  ops/s
[info] JsonFramingBenchmark.counting_offer_5        thrpt   30  1458837,532 ± 30467,879  ops/s
```

Now the really weird part is when the next level of JIT kicks in (example on counting_offer_5):
```text
[info] # Warmup Iteration   1: 2510990,608 ops/s
[info] # Warmup Iteration   2: 2526441,393 ops/s
[info] # Warmup Iteration   3: 2641141,125 ops/s
[info] # Warmup Iteration   4: 2659889,369 ops/s
[info] # Warmup Iteration   5: 2737527,462 ops/s
[info] # Warmup Iteration   6: 2638352,978 ops/s
[info] # Warmup Iteration   7: 2551206,206 ops/s
[info] # Warmup Iteration   8: 2791854,785 ops/s
[info] # Warmup Iteration   9: 2771254,012 ops/s
[info] # Warmup Iteration  10: 2774640,490 ops/s
[info] # Warmup Iteration  11: 2714039,178 ops/s
[info] # Warmup Iteration  12: 2623812,107 ops/s
[info] # Warmup Iteration  13: 2631459,200 ops/s
```
so far so good. And then:

```text
[info] # Warmup Iteration  14: 1625617,941 ops/s
[info] # Warmup Iteration  15: 1521581,545 ops/s
[info] # Warmup Iteration  16: 1487617,934 ops/s
[info] # Warmup Iteration  17: 1523624,664 ops/s
[info] # Warmup Iteration  18: 1455953,394 ops/s
[info] # Warmup Iteration  19: 1433586,772 ops/s
[info] # Warmup Iteration  20: 1435661,055 ops/s
[info] # Warmup Iteration  21: 1409643,865 ops/s
[info] # Warmup Iteration  22: 1405667,711 ops/s
[info] # Warmup Iteration  23: 1421820,338 ops/s
[info] # Warmup Iteration  24: 1436581,242 ops/s
[info] # Warmup Iteration  25: 1456347,914 ops/s
[info] # Warmup Iteration  26: 1502848,906 ops/s
[info] # Warmup Iteration  27: 1544719,316 ops/s
[info] # Warmup Iteration  28: 1497289,444 ops/s
[info] # Warmup Iteration  29: 1369107,820 ops/s
[info] # Warmup Iteration  30: 1375654,576 ops/s
[info] Iteration   1: 1401525,462 ops/s
[info] Iteration   2: 1400224,279 ops/s
[info] Iteration   3: 1409271,647 ops/s
[info] Iteration   4: 1456632,194 ops/s
[info] Iteration   5: 1461690,497 ops/s
[info] Iteration   6: 1467002,056 ops/s
[info] Iteration   7: 1436177,597 ops/s
[info] Iteration   8: 1432764,005 ops/s
[info] Iteration   9: 1502544,198 ops/s
```
(this also happens on `counting_1`)

There is no such cliff on master. 
`VM version: JDK 1.8.0_171, OpenJDK 64-Bit Server VM, 25.171-b11`

Clearly something very wrong seems to happen at some stage of JITing… any idea?

Commit:
Removing patmat and attempting to use more tailcalls

Commit:
Attempting to help the JIT find that each state is its own inline/optim domain

Commit:
Experimented with deep-stack of seekNextEvent → useless.

Commit:
Experimenting with interning the buffer within the inner loop machinery

Commit:
Remove the ParserStateMachinery machinery: the JVM does it automatically

Commit:
Attempted to hand-specialize for ByteString1C: useless.

Commit:
Revert "Attempted to hand-specialize for ByteString1C: useless."

This reverts commit 332af501e7b16327abb564504e6b86f24eac6e7e.

Commit:
New benchmark: what if we receive objects by chunks of 5 bytes?

Commit:
Rewrite the state machine to remove two tracking variables

we encode whether we're currently before or within the object as a
property of the current state, which removes the need to advance
trimFront or charsInObject together with "pos".

Meanwhile we ensure we never compact the ByteStrings as we gain no speed.

Commit:
removing commented out debug printlns

cchepelov commented:
So it's been a trip through *quite* a rabbit hole, and today's revision still exhibits the "cliff" after so many iterations. Nevertheless, it has interesting speed properties:

**19OCT18 11:55 @ -wi 3 -i 3** 
```text
Benchmark                                             Mode  Cnt        Score         Error  Units
JsonFramingBenchmark.counting_1                      thrpt    3  4103269,687 ±  512715,754  ops/s
JsonFramingBenchmark.counting_long_document          thrpt    3      608,480 ±      77,174  ops/s
JsonFramingBenchmark.counting_offer_5                thrpt    3  3783312,469 ± 1163705,491  ops/s
JsonFramingBenchmark.counting_offer_5by5bytes        thrpt    3   635380,540 ±   31701,942  ops/s
JsonFramingBenchmark.counting_offer_5x5              thrpt    3  2584182,709 ±  274134,668  ops/s
```

**19OCT18 12:00 @ -wi 50 -i 10** 
```text
Benchmark                                            Mode  Cnt        Score       Error  Units
JsonFramingBenchmark.counting_1                     thrpt   10  3375454,452 ± 33033,390  ops/s
JsonFramingBenchmark.counting_long_document         thrpt   10      616,481 ±     6,076  ops/s
JsonFramingBenchmark.counting_offer_5               thrpt   10  2112254,668 ± 16235,293  ops/s
JsonFramingBenchmark.counting_offer_5by5bytes       thrpt   10   591742,440 ±  6086,794  ops/s
JsonFramingBenchmark.counting_offer_5x5             thrpt   10  2335804,030 ± 25440,370  ops/s
```

**MASTER + new benchmarks @ -wi 3 -i 3**
```text
Benchmark                                             Mode  Cnt        Score        Error  Units
JsonFramingBenchmark.counting_1                      thrpt    3  3004525,601 ± 117561,449  ops/s
JsonFramingBenchmark.counting_long_document          thrpt    3      223,084 ±     13,730  ops/s
JsonFramingBenchmark.counting_offer_5                thrpt    3  2571295,037 ± 214324,542  ops/s
JsonFramingBenchmark.counting_offer_5by5bytes        thrpt    3   515932,275 ± 124061,419  ops/s
JsonFramingBenchmark.counting_offer_5x5              thrpt    3  1701273,433 ±  51550,842  ops/s
```

**MASTER + new benchmarks @ -wi 50 -i 10**
```text
Benchmark                                             Mode  Cnt        Score       Error  Units
JsonFramingBenchmark.counting_1                      thrpt   10  2957010,988 ± 19847,931  ops/s
JsonFramingBenchmark.counting_long_document          thrpt   10      221,140 ±     1,046  ops/s
JsonFramingBenchmark.counting_offer_5                thrpt   10  2584635,725 ±  9334,813  ops/s
JsonFramingBenchmark.counting_offer_5by5bytes        thrpt   10   540532,705 ±  2990,755  ops/s
JsonFramingBenchmark.counting_offer_5x5              thrpt   10  1619520,343 ±  7315,897  ops/s
``` 
The two new benchmarks exhibit what is happening under two sub-scenarios:
 - `counting_offer_5x5` is when there are large offers (5 offers of a string containing 5 objects each) before downstream gets around to polling
 - `counting_offer_5by5bytes` is when the offer (from the original "5 objects string") drip by chunks of 5 bytes, something that could happen due to networking conditions.

in a nutshell:
- the revised code is significantly faster on a low-ish iteration count (before the 40 millionth operation or so, it delivers betwen +30% and +300% more itertions per second
- with the exception of the `counting_offer_5` benchmark (where one call to `.offer` provides 5 contiguous JSON objects within a single ByteString, a scenario which isn't unlikely at all under regular network conditions), the revised code is always faster even after the apparent "cliff"
- I was unable to see anything different happening even while squinting through the output of `
jmh:run -wi 3 -i 3 -f1 -t1 -jvmArgsPrepend "-XX:+PrintCompilation -XX:+UnlockDiagnosticVMOptions -XX:+PrintInlining" -prof perf .*JsonFraming.*
` for various variations of `-wi 3 -i 3`. In particular:
  -  the "hottest region" code appears identical with short or long warm-up
  - running the benchmark on a 6th generation Intel machine (rather than this 7th) exhibits no such cliff
  - the only JIT traffic happening in the vicinity of the cliff seems related to a ConcurrentHashMap and classloading activity (of which there is none within the ```JsonObjectParser``` benchmark itself)

Commit:
JsonObjectParser$.isWhitespace is private, thanks Mima

johanandren commented on akka-stream/src/main/mima-filters/2.5.x.backwards.excludes:
```diff
@@ -60,4 +60,14 @@ ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowO
+ProblemFilters.exclude[IncompatibleResultTypeProblem]("akka.stream.impl.JsonObjectParser.SquareBraceEnd")
+ProblemFilters.exclude[IncompatibleResultTypeProblem]("akka.stream.impl.JsonObjectParser.CurlyBraceStart")
+ProblemFilters.exclude[IncompatibleResultTypeProblem]("akka.stream.impl.JsonObjectParser.SquareBraceStart")
+ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.impl.JsonObjectParser.isWhitespace")
```
This is in the wrong file, should be `2.5.18.backwards.excludes` (this won't make it into 2.5.18 which is in the process of being released and the actual version number is the latest released)

cchepelov replied:
pushing a correction for this, thanks!

johanandren reviewed (approved):
LGTM except the wrong mima file

Commit:
rebasing MIMA filters in their proper place

Commit:
avoid overriding def toString() to return the standard result anyway

Commit:
don't hide InitialState; get rid of UnknownState (unused)

jrudolph commented:
>     * the revised code is always faster even after the apparent "cliff"

The most simple cause for such a cliff with current (Intel) processors is thermal throttling. That is, turbo mode can only be sustained for short while (e.g. 10s), after which the CPU is throttled down to nominal frequency.

Try running `i7z` in another window to see that in action. I always use `echo "1" | sudo tee /sys/devices/system/cpu/intel_pstate/no_turbo` on Linux to make sure turbo is disabled when running benchmarks.

Even with that setting, you still need to keep an eye on the frequency while running the benchmark because thermal throttling can also happen when the CPU gets to hot with nominal frequencies (dirty fan, bad laptop case design, problems with heat dissipation etc.). This can differ a lot between machines. Afaiu, recent Intel U-suffix CPU may be more affected because they offer burst speeds similar to previous CPU but with 1/3 of the TDP which means they need to keep a tight thermal regime and throttle aggressively under sustained load.

raboof commented:
I'm afraid JsonObjectParser has been quite substantially changed since this PR :(. I've tried rebasing for a bit but I'm not deep enough in this code to judge what is reasonable.

It pains me to close such a nice PR after such a long time, but for now I don't see a reasonable other way forward. Of course if your improvements still make sense on master it would be great if you could update this PR (or open up a new one) - I promise we'll try to be more responsive then...
